File: go/engine_conformance_test.go
// Copyright (c) 2026 Richard Rodger and other contributors, MIT License

package tabnasmarkdown

// engine_conformance_test.go — engine-path conformance: the full CommonMark
// 0.31.2 suite (652 examples, gfm:false) and the GFM extension corpus (24
// examples, gfm:true) run through Make(...).ParseMeta(...) — the engine
// lexing lines, the rules dispatching them — with byte-for-byte HTML
// comparison and an AST comparison against the engine-free path.
//
// commonmark_test.go asserts the same corpus over the engine-free modules;
// this suite is the other leg of the dual-path contract (dx-report §42):
// the conformance claim holds on the code path the plugin actually runs,
// not just on the reference implementation. The native tree comes back via
// meta["md"]["keepTree"] and is rendered with the same RenderHTML the
// engine-free path uses, so the comparison isolates the parse — a
// difference here is a parsing difference, never a rendering one. Mirrors
// ts/test/engine-conformance.test.ts.

import (
	"reflect"
	"testing"

	parser "github.com/tabnas/parser/go"
)

func checkEngineExample(t *testing.T, j *parser.Tabnas, c specCase, opts Options) {
	t.Helper()

	meta := map[string]any{"md": map[string]any{"keepTree": true}}
	ast, err := j.ParseMeta(c.Markdown, meta)
	if err != nil {
		t.Fatalf("example %d: engine parse error: %v", c.Example, err)
	}

	tree, _ := meta["md"].(map[string]any)["tree"].(*MdNode)
	if tree == nil {
		t.Fatalf("example %d: keepTree returned no native tree", c.Example)
	}

	if got := RenderHTML(tree, Options{GFM: tree.GFM}); got != c.HTML {
		t.Errorf("example %d: engine-path HTML\nmarkdown: %q\n     got: %q\n    want: %q",
			c.Example, c.Markdown, got, c.HTML)
	}

	direct := ParseDocument(c.Markdown, opts)
	if !reflect.DeepEqual(jsonFlatten(ast), jsonFlatten(direct)) {
		t.Errorf("example %d: engine-path AST diverges from engine-free path\nmarkdown: %q",
			c.Example, c.Markdown)
	}
}

func TestEngineCommonMarkSpec(t *testing.T) {
	cases := loadSpecCases(t)
	j := Make(map[string]any{"gfm": false})
	opts := ResolveOptions(map[string]any{"gfm": false})
	for _, c := range cases {
		checkEngineExample(t, j, c, opts)
	}
}

func TestEngineGFMSpec(t *testing.T) {
	cases := loadGFMCases(t)
	j := Make()
	opts := ResolveOptions(nil)
	for _, c := range cases {
		checkEngineExample(t, j, c, opts)
	}
}

File: rs/tests/engine_conformance_test.rs
// Engine-path conformance: the full CommonMark 0.31.2 suite (652
// examples, gfm:false) and the GFM extension corpus (24 examples,
// gfm:true) run through `make_with(..)` and `parse_keep_tree` (the engine
// lexing lines, the rules dispatching them) with byte-for-byte HTML
// comparison and an AST comparison against the engine-free path.
//
// commonmark_test.rs asserts the same corpus over the engine-free
// modules; this suite is the other leg of the dual-path contract
// (dx-report §42): the conformance claim holds on the code path the
// plugin actually runs, not just on the reference implementation. The
// native tree comes back via the `keepTree` handshake and is rendered
// with the same `render_html` the engine-free path uses, so the
// comparison isolates the parse: a difference here is a parsing
// difference, never a rendering one. Mirrors
// ts/test/engine-conformance.test.ts and go/engine_conformance_test.go.

mod common;

use common::{load_gfm_cases, load_spec_cases, to_json, SpecCase, CM_OPTS, GFM_OPTS};
use tabnas::Tabnas;
use tabnas_markdown::{make_with, parse_document, parse_keep_tree, render_html, Options};

fn check_engine_example(tn: &Tabnas, c: &SpecCase, opts: &Options, failures: &mut Vec<String>) {
    let (ast, tree) = match parse_keep_tree(tn, &c.markdown) {
        Ok(result) => result,
        Err(error) => {
            failures.push(format!(
                "example {}: engine parse error: {error}",
                c.example
            ));
            return;
        }
    };
    let Some(tree) = tree else {
        failures.push(format!(
            "example {}: keepTree returned no native tree",
            c.example
        ));
        return;
    };

    let got = render_html(&tree, None);
    if got != c.html {
        failures.push(format!(
            "example {}: engine-path HTML\nmarkdown: {:?}\n     got: {:?}\n    want: {:?}",
            c.example, c.markdown, got, c.html
        ));
    }

    let direct = parse_document(&c.markdown, opts);
    if to_json(&ast) != to_json(&direct) {
        failures.push(format!(
            "example {}: engine-path AST diverges from engine-free path\nmarkdown: {:?}",
            c.example, c.markdown
        ));
    }
}

#[test]
fn engine_commonmark_spec() {
    let cases = load_spec_cases();
    let tn = make_with(&CM_OPTS);
    let mut failures = Vec::new();
    for c in &cases {
        check_engine_example(&tn, c, &CM_OPTS, &mut failures);
    }
    assert!(
        failures.is_empty(),
        "{} failures:\n{}",
        failures.len(),
        failures.join("\n")
    );
}

#[test]
fn engine_gfm_spec() {
    let cases = load_gfm_cases();
    let tn = make_with(&GFM_OPTS);
    let mut failures = Vec::new();
    for c in &cases {
        check_engine_example(&tn, c, &GFM_OPTS, &mut failures);
    }
    assert!(
        failures.is_empty(),
        "{} failures:\n{}",
        failures.len(),
        failures.join("\n")
    );
}

/// The handshake itself: a tree comes back only when asked for, never
/// for an empty source, and never stale.
#[test]
fn keep_tree_handshake() {
    let tn = make_with(&GFM_OPTS);

    let (ast, tree) = parse_keep_tree(&tn, "# hi\n").expect("parses");
    assert_eq!(to_json(&ast)["children"][0]["type"], "heading");
    let tree = tree.expect("a tree for a non-empty source");
    assert_eq!(render_html(&tree, None), "<h1>hi</h1>\n");

    let (ast, tree) = parse_keep_tree(&tn, "").expect("parses");
    assert_eq!(
        to_json(&ast),
        serde_json::json!({"type": "document", "children": []})
    );
    assert!(tree.is_none(), "the engine short-circuits an empty source");

    // A plain parse parks nothing.
    tn.parse("x\n").expect("parses");
    let (_, tree) = parse_keep_tree(&tn, "").expect("parses");
    assert!(tree.is_none());
}
